qp_deblur/deblur.py

```diff
@@ -157,6 +158,145 @@ def generate_sepp_placements(seqs, out_dir, threads, reference_phylogeny=None,
+
+    # create a dummy sequence input file
+    file_input = '%s/input.fasta' % out_dir
+    with open(file_input, 'w') as f:
+        f.write('>dummySeq\n')
+        f.write('TACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGATGGA'
+                'CAAGTCTGATGTGAAAGGCTGGGGCCCAACCCCGGGACTGCATTGGAAACTGCCCGTCTT'
+                'GAGTG\n')
+    std_out, std_err, return_value = system_call(
+        'cd %s; run-sepp.sh %s dummy -x 1 -a %s -t %s' %
+        (out_dir, file_input, file_reference_alignment,
+         file_reference_phylogeny))
+    if return_value != 0:
+        error_msg = ("Error running SEPP:\nStd out: %s\nStd err: %s"
+                     % (std_out, std_err))
+        raise ValueError(error_msg)
+
+    # take resulting placement.json and turn it into the template by
+    # clearing the list of placements
+    file_template = '%s/tmpl_dummy_placement.json' % out_dir
+    with open('%s/dummy_placement.json' % out_dir, 'r') as f:
+        placements = json.loads(f.read())
+        placements['placements'] = []
+        with open(file_template, 'w') as fw:
+            json.dump(placements, fw)
+
+    return (file_template, '%s/dummy_rename-json.py' % out_dir)
+
+
+def generate_insertion_trees(placements, out_dir,
+                             reference_template=None,
+                             reference_rename=None):
+    """Generates phylogenetic trees by inserting placements into a reference
+
+    Parameters
+    ----------
+    placements : dict of strings
+        keys are the seqs, values are the new placements as JSON strings
+    out_dir : str
+        The job output directory
+    reference_template : str, optional
+        Filepath to the reference placement json file.
+        This file can be produced via _generate_template_rename() and should be
+        stored in the plugin package, because it can re used.
+        If None, it falls back to the Greengenes 13.8 99% reference.
+    reference_rename : str, optional
+        Similar to reference_template, but a filepath to the generated python
+        renaming script to undo the name scaping post guppy.
+        If None, it falls back to the Greengenes 13.8 99% reference.
+
+    Returns
+    -------
+    str
+        The filepath of the phylogenetic insertion tree in Newick format.
+    """
+
+    # create a valid placement.json file as input for guppy
+    file_ref_template = resource_filename(
+        Requirement.parse('qp_deblur'),
+        'qp_deblur/assets/tmpl_gg13.8-99_placement.json')
+    if reference_template is not None:
+        file_ref_template = reference_template
+    if not exists(file_ref_template):
+        raise ValueError("Reference template '%s' does not exits!" %
+                         file_ref_template)
+    with open(file_ref_template, 'r') as f:
+        plcmnts = json.loads(f.read())
+        for sequence, placement in placements.items():
+            plcmnts['placements'].append({'p': placement,
```

Comment:
Can you use extend with a list comprehension instead of append? It is faster and given the number of items that placements can have it may be significant:
```python
In [10]: l = list([str(i), str(i+1)] for i in range(100))

In [11]: def f1(l):
    ...:     pl = []
    ...:     for a, b in l:
    ...:         pl.append({'p': a, 'nm': b})
    ...:         

In [12]: def f2(l):
    ...:     pl = []
    ...:     pl.extend([{'p': a, 'nm': b} for a, b in l])
    ...:     

In [13]: %timeit f1(l)
100000 loops, best of 3: 20.6 µs per loop

In [14]: %timeit f2(l)
100000 loops, best of 3: 15.8 µs per loop
```

Reply:
Cool. I did not know that. Changed.
I figure reading the placements file from disk will dominate runtime for this function, but it is always good to keep the execution time low.